### PR TITLE
Add s6 finish script to ensure graceful termination of rtorrent 

### PR DESCRIPTION
**Pull request by deed02392**
…when stopping container.

Addresses issue #58

### CODE REVIEW

**Comment by sparklyballs** [comment]
not convinced of a need for this as the process manager already sends a kill signal

plus arbitrary sleep times make no account for slow or heavily loaded systems and on those would result in the pid file remaining unreaped

**Comment by deed02392** [comment]
@sparklyballs why are you being a dick?

First you close issue #57 because you claimed my permissions were wrong - they weren't.
Then you close #58 saying it's not an issue, when it clearly is if you'd bother to do testing of it.
And now you've closed this PR saying it's not necessary, again indicating you did not read or test issue #58 before closing it. If it's just the usage of sleep that bothers you, I can update the PR.

Again - s6-nuke sends SIGTERM to rtorrent which kills it ungracefully.

As described in the commit, 5 seconds is not arbitrary, but the time that rtorrent gives trackers to respond to the leave announcement before it closes files and unlinks the lock file.

**Comment by sparklyballs** [comment]
`@sparklyballs why are you being a dick?`

really no need for that, we're not obligated to accept a PR.

if you want this operation for your personal use case, then volume mount the script as file only into the location

**Comment by deed02392** [comment]
@sparklyballs I appreciate you're not obligated, but it feels disrespectful to not even read my comments or commit messages before closing them dismissively. This is an actual fix.

**Comment by horjulf** [comment]
This is an actual issue, why wasn't this merged or at least an alternative method discussed ?
The SIGINT can be sent and the pid monitored until rtorrent exits and the container stopped.

**Comment by deed02392** [comment]
@f-santos I know. See issue #58 which describes the problem. Apparently @sparklyballs doesn't care that rtorrent is killed ungracefully, or can't be bothered to read.

**Comment by sparklyballs** [comment]
your sniping is getting somewhat irksome 

i read the PR , how else would i know about a spurious sleep ? 

for your information this does not actually work without further configuring and then after that there is no way of setting a time beyond which slow systems would have time to respond which is why i rejected the PR in the first place had you been bothered to read my comment.

**Comment by deed02392** [comment]
@sparklyballs "not convinced of a need for this as the process manager already sends a kill signal"

My response:
"s6-nuke sends SIGTERM to rtorrent which kills it ungracefully."

So I explained this is the fundamental issue which you don't seem to be acknowledging - do you agree it is a problem or not?

If you are concerned about this not succeeding on slow systems, then you are acknowledging it's an issue, yet you have closed the issue with the reason it is to do with NFS, which is not accurate whatsoever.

**Comment by sparklyballs** [comment]
no, saying the "fix" will not work on slower systems is more of an acknowledgement that the "fix" does not work and does nothing to say whether it is required or not, it merely says it wouldn't work anyway.

**Comment by deed02392** [comment]
OK, so it's fine that rtorrent is ungracefully terminated causing a spurious lock file that you have to delete as part of your startup scripts. Thanks. Maybe @CHBMB, @IronicBadger, @JAGFin1 and @tobbenb can agree with you and then I will finally rest my case.

**Comment by CHBMB** [comment]
@deed02392 I don't know enough to comment on the s6 issue, however I don't approve of you referring to @sparklyballs as a dick.  I've been subject to similar name calling on github and there's absolutely no call for it and I suggest you work on your interpersonal skills.

**Comment by deed02392** [comment]
To be fair @CHBMB, he was being a dick when closing my issues and PRs without any constructive discussion. I think it's he that needs to work on his interpersonal skills, frankly. The only reason I am persisting here despite the obvious frustration it is causing me is because I think your selection of containers is one of the best out there.

You have another user here agreeing with me now on the issue that sparklyballs closed, that he still has not demonstrated an understanding of. 

I will consider your suggestion and in the meantime, suggest to you that you take an impartial look at the discussion here and decide whether there aren't two parties at fault and then think about how we can move forward productively.

**Comment by sparklyballs** [comment]
you accused me of not reading your PR when i clearly had as how the hell did you think i knew about the sleep command ?

then your first response was to call me a dick. 

you have taken enough of my time and i will not be wasting any more of it with you since you once again have resorted to childish insults when you think you aren't getting your way.

**Comment by deed02392** [comment]
It was hardly my first response! I regret allowing myself to get frustrated to the extent I resorted to name calling, but if you actually look back at our initial discussion you'll notice I was being entirely police and gracious, only to have you keep curtly dismissing my input.

I even attempted to reach out to you directly on IRC but you didn't return my message.

Seriously, I just want to help you guys out. I can run off my own branch no problem for now. I'll stop messaging here now. Thanks for your work in getting this project as far as it is. Peace.

**Comment by CHBMB** [comment]
@deed02392 yes he closed your PR, he gave a reason, he wasn't rude.  Your first reply to that was by calling him a dick.  Now you got every right to feel however you want, but you're not doing yourself any favours by once again referring to him as a dick again.

I got no problems discussing the issue, 

I have a MASSIVE issue with you twice referring to @sparklyballs as a dick.  He hasn't resorted to calling you names in return so I think my suggestion that you need to work on your interpersonal skills is valid.

In no situation like this is name calling reasonable, and it will not be tolerated.  

I know that behind the scenes @aptalca and @sparklyballs have been working on shutdown stuff similar to this and due to limitations of docker have found several issues.

The general impression I'm getting is that your PR would replace one issue with another, now that 2nd issue may not affect you or even the other user that commented, but that doesn't mean it's not an issue, just that it doesn't affect 2 people.

**Comment by horjulf** [comment]
Please, can we just technically discuss the issue here.

From what I see, the default docker stop timeout for containers is 10 seconds so that gives time for trying to stop rtorrent gracefully, plus the default timeout can be overridden by container when its created (--stop-timeout).

I suggest to give the user the choice of how long to wait for rtorrent to close or just leave the default 10 seconds, either way on worst case scenario in a slow system you get the current behaviour since the container is killed by the docker stop timeout.

**Comment by deed02392** [comment]
@f-santos that's not exactly correct. The problem is the default behaviour of this container, upon receiving stop, is that it attempts to stop all running processes with the `s6-nuke` program, which sends a signal to rtorrent telling it to terminate immediately (ungracefully). This leaves behind its lock file and puts session data at risk of not syncing and therefore corrupting ongoing torrents.

So it's not just that the default timeout of 10 seconds may not be long enough, it's also a coding issue in the existing container.

This PR could be improved by simply monitoring the rtorrent process after sending the appropriate signal. There will be a delay of _at least_ 5 seconds because rtorrent waits that long for trackers to respond to the 'close' announces. That is the reason for the 5 second sleep - I then left it up to the rest of the `s6-nuke` script to be more aggressive.

After those 5 seconds, all rtorrent has to do is close its file handles, which shouldn't take longer than s6-nuke is willing to wait, but I agree for those with hundreds or thousands of torrents, a customisable delay would be good.

**Comment by aptalca** [comment]
That is not quite correct. S6 actually sends a sigterm to running processes to stop gracefully. The issue is that there is a timeout, which currently is 3 seconds. Then it sends a sigkill.

The 5 second wait for trackers you mention, is that hardcoded or is that what you observe in your system?  Can you point to a reference? If it's hardcoded, we could increase the timeout from 3 to 6 seconds and it may solve the issue. But if it's just what you observe, then it wouldn't work on slower systems. We would have to figure out something else.

**Comment by deed02392** [comment]
I never disputed that the signal `s6-nuke` sent was something other than SIGTERM. I said the signal it sent was the wrong one, and it is.

Again, the problem is that SIGTERM causes an ungraceful shutdown.

https://github.com/rakshasa/rtorrent/wiki/User-Guide
On the wiki it says the behaviour on receiving sigterm is 'shut down immediately'. That's the same as ungracefully, which I deem to be so because when I tested this, the lock file was left behind on receiving a sigterm.

The 5 second wait is again listed right there on that user guide. It is hard coded: https://github.com/rakshasa/rtorrent/blob/ce0008510723b6c96c86f8fe944affc84f90cd69/src/control.cc#L181

I do not think increasing the timeout will work though, as I said, because rtorrent needs SIGINT to shut down normally, not SIGTERM.

This PR tells rtorrent to shut down normally, then waits the same hard coded time for it to do so. I believe rtorrent then has however long it takes for SIGTERM to be sent, that other defined delay which you say is 3 seconds, then SIGKILL is sent. rtorrent will ignore SIGTERM since it's already shutting down, so it basically has 3 seconds AFTER the hard coded 5 seconds to close files, which as I said above should be long enough, and if it isn't for people with what will need to be either thousands of torrents or poorly specced/failing storage, then we can do it such that we wait for rtorrent's PID to disappear before resuming the shutdown process. Obviously that has its own risks if rtorrent fails to shut down properly, which is why I didn't implement that.

**Comment by aptalca** [comment]
I just took a look at it and the issue is entirely different. The s6 run process starts a screen session, which spawns the rtorrent process, as well as starting the logging process. When shutting down, s6 sends the SIGTERM to the screen process and not rtorrent, which doesn't do anything. As a result, rtorrent process gets a SIGKILL after 3 seconds.

By the way, SIGTERM to the rtorrent process results in a clean shutdown, with the lock file removed (although without the 5 second wait for confirmation from the trackers).

We'll fix this by separating the rtorrent and the logging processes so the rtorrent process gets the SIGTERM properly during shutdown.

**Comment by deed02392** [comment]
Don't we still want to perform a normal shut down of rtorrent with sigint?

**Comment by sparklyballs** [comment]
there is no issue to start with and nothing to "fix"

every option to start rtorrent [here](https://wiki.archlinux.org/index.php/RTorrent) includes a routine to cull the lock file

introducing something to interfere with the shutdown routine not only doesn't take into account slow systems it introduces a potential race condition that could leave open screen sessions that either consume resources or leave a fail condition due to open session and then requiring a routine to cull those on startup

so in summary "fixing" something that doesn't actually require fixing introduces more problems than solving anything that it doesn't actually need to solve

**Comment by aptalca** [comment]
With regards to sigterm and sigint, the only difference is the announce to trackers,which I don't care about. Neither do the package scripts @sparklyballs linked to above.

Also as @sparklyballs mentioned above, I haven't seen any corruption issues due to unclean shutdown apart from when combined with mounted shares, which could have been due to mounted shares alone. Docker does not play well with them.

In this case, having to run rtorrent in screen presents a big problem with s6 because s6 is a "process" manager, not a "service" manager. When you have a process (screen) spawning another (rtorrent) it doesn't work properly with s6 in terms of shutting down. 

With that said, although I'm not convinced that there is a data corruption issue, I realize that there is unclean shutdown of rtorrent (undesirable but not necessarily an issue). I'll try a couple of methods to circumvent the s6 shortcoming, but no promises.

Also, this pr would not work because the cont-finish script runs before the s6 processes are attempted to be brought down, meaning when the rtorrent process is killed, s6 will relaunch it.

**Comment by deed02392** [comment]
I've not experienced any corruption yet either, but I perceive it to be a risk.

In the linked package scripts, there are 2 unit files listed. I'm not sure why you said they don't care about SIGTERM vs SIGINT, because if you look they both use killall with `-s 2` which sends SIGINT.

This PR 'works for me'. rtorrent does not get relaunched during shutdown of the container with this PR.
